runtime/compiler/trj9/runtime/RelocationRuntime.cpp

```diff
@@ -657,6 +657,34 @@ TR_RelocationRuntime::relocateMethodMetaData(UDATA codeRelocationAmount, UDATA d
       _exceptionTable->riData = (void *) (((U_8 *)_exceptionTable->riData) + dataRelocationAmount);
       }
 
+   if (_comp->getOptions()->enableDebugCounters())
+      {
+      _exceptionTable->debugSlot1 = (void *) (((U_8 *)_exceptionTable->debugSlot1) + dataRelocationAmount);
+
+      uint8_t *cursor = (uint8_t *)_exceptionTable->debugSlot1;
+
+      uint32_t numDebugCounterNames = *(uint32_t *)cursor;
+      cursor += sizeof(uint32_t);
+      for (uint32_t i = 0; i < numDebugCounterNames; i++)
+         {
+         const char *key = (const char *)(*(uintptrj_t *)cursor);
+         cursor += sizeof(uintptrj_t);
+
+         uint32_t size = *(uint32_t *)cursor;
+         cursor += sizeof(uint32_t);
+
+         if (!_comp->getCounterName(key))
+            {
+            // Need to allocate persistent memory since the metadata can be reclaimed
```

Comment:
> Is it reclaimed or just disclaimed?

The metadata that gets allocated in the data cache gets reclaimed, ie, freed back into the data cache for reuse of something else. 

> the entire shared classes cache should be mapped into the address space, so I would have guessed you could just use pointers to the names stored in the meta data.
>
> In that case, it would be better to store the references as offsets from the beginning of the shared classes cache, so then you can easily convert the offset in the relocation data into a pointer (to a spot in the meta data holding the name) by just adding the cache address (we do that in a bunch of other places e.g. class chains).

At AOT compile, when we store all the relocation records to the buffer, the metadata hasn't be created yet (I believe), so the relocation record can't know the offset at which the string will be stored. That's why I used the string ptr as both the key and the value during the AOT compile step.

However, what you're suggesting does work at AOT load time. since I can calculate the offset into the memory mapped shared cache from the metadata that gets memcpy'd into data cache allocated memory. That way I don't need to allocate the strings into persistent memory.

All that said, I just realized that there's a problem with the current implementation. `TR::DebugCounterGroup::findCounter` finds the counter from a hashtable that's hashed via a `const char *`. For counters that are created during a compile where the name is dynamically generated, my current implementation is fine. However, for counters who's names are known at compile time (ie stored in the BSO section of the dll), my implementation breaks down; this is because if the counter is enabled for multiple methods, the same string will get stored in multiple metadatas, and each of those will result in a new counter created at relo time (because they will have a different `const char *` ptr as the key in the counter hash). I don't think it's a massive functional problem, the accumulation will just not be truly accumulative and instead will result in the developer having to add up stuff manually which is annoying. 

I'm gonna have to think about how to get around this issue without dealing directly with the SCC; I'm reluctant to deal with the SCC directly because, as far as I could tell, the only way to store data to it was to hash it with a key, and generating a unique key involved (at least looking at how classes in the class chain are stored) taking the pointer and converting it to an ASCII string. If there's an easier way to deal with the SCC I'm more than happy to look into it.